From 982f86242e088ea71ec6aff077bb856c036952fb Mon Sep 17 00:00:00 2001
From: Martin Jansa <martin2.jansa@lgepartner.com>
Date: Sun, 10 Sep 2023 15:20:44 +0000
Subject: [PATCH] Revert "qemu/osdep: Remove fallback for MAP_FIXED_NOREPLACE"

:Release Notes:
This reverts commit c42e77a90d9244c8caf76fe0e54f84200430a4e1.
For qemu-8.1 we need newer libc than 2.27 used in ubuntu-18.04 to
support MAP_FIXED_NOREPLACE after qemu removed fallback for it in:
https://github.com/qemu/qemu/commit/c42e77a90d9244c8caf76fe0e54f84200430a4e1
but after switching from buildtools to buildtools-extended which
include gcc and libc the gdk-pixbuf-native started to fail.

:Detailed Notes:
gdk-pixbuf-native issue can be avoided by patching glib-2.0 not to
use close_range even with newer glib as shown in:
http://gpro.lge.com/c/webos-pro/meta-lg-webos/+/370231 glib-2.0-native: gspawn: avoid close_range

But that's not all there are more failures caused by buildtools-extended
in our multilib builds (because buildtools-extended doesn't include
lib32-glibc and some of our components pass -m32 in native builds e.g.:

1) lib32-ntllib-11.5.1-r0 do_compile
   meta-webos-pro/recipes-webos/ntllib/ntllib_11.5.1.bb
   HOST_CXX_append = "${@ '' if '${TARGET_ARCH}' == 'aarch64' or '${TARGET_ARCH}' == 'x86_64' else '-m32'}"

   causing http://gecko.lge.com:8000/Errors/Details/690072

   TOPDIR/buildtools-4.2.3/sysroots/x86_64-pokysdk-linux/usr/include/gnu/stubs.h:7:11: fatal error: gnu/stubs-32.h: No such file or directory
    7 | # include <gnu/stubs-32.h>
      |           ^~~~~~~~~~~~~~~~

2) lib32-luajit-2.1.0~beta3-220721-r0 do_compile
   http://gecko.lge.com:8000/Errors/Details/690073

   TOPDIR/buildtools-4.2.3/sysroots/x86_64-pokysdk-linux/usr/include/gnu/stubs.h:7:11: fatal error: gnu/stubs-32.h: No such file or directory
    7 | # include <gnu/stubs-32.h>
      |           ^~~~~~~~~~~~~~~~

So it might be better to get back to revert qemu commit to remove fallback
and continue to use regular buildtools until SWP infra is updated to
supported OS.

:Testing Performed:
Only build tested.

:QA Notes:
No change to image.

:Issues Addressed:
[WRP-13301] Create GPVB with Yocto 4.3 Nanbield
---
Upstream-Status: Inappropriate [LGE specific due to old libc in ubuntu 18.04 on LGE servers]

 include/qemu/osdep.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/qemu/osdep.h b/include/qemu/osdep.h
index 21ef8f1699..cc61b00ba9 100644
--- a/include/qemu/osdep.h
+++ b/include/qemu/osdep.h
@@ -289,6 +289,9 @@ void QEMU_ERROR("code path is reachable")
 #ifndef MAP_ANONYMOUS
 #define MAP_ANONYMOUS MAP_ANON
 #endif
+#ifndef MAP_FIXED_NOREPLACE
+#define MAP_FIXED_NOREPLACE 0
+#endif
 #ifndef MAP_NORESERVE
 #define MAP_NORESERVE 0
 #endif
